File: dto-mapping-benchmark/src/main/java/com/example/dtomappingbenchmark/v4/mapper/UserMapper.java
package com.example.dtomappingbenchmark.v4.mapper;

import com.example.dtomappingbenchmark.v4.entity.User;
import com.example.dtomappingbenchmark.v4.dto.UserDto;

import java.util.ArrayList;

public class UserMapper {
    public static User toEntity(UserDto dto){
        return User.builder()
                .id(dto.getId())
                .name(dto.getName())
                .email(dto.getEmail())
                .posts(new ArrayList<>())
                .build();
    }

    public static UserDto toDto(User entity){
        return UserDto.builder()
                .id(entity.getId())
                .name(entity.getName())
                .email(entity.getEmail())
                .postList(entity.getPosts().stream()
                        .map(PostMapper::toDto)
                        .toList())
                .build();
    }
}

File: object-mapping-benchmark-v5/src/main/kotlin/yu/selab/cj/objectmappingbenchmarkv5/mapper/UserMapper.kt
package yu.selab.cj.objectmappingbenchmarkv5.mapper

import org.mapstruct.Mapper
import org.mapstruct.Mapping
import yu.selab.cj.objectmappingbenchmarkv5.user.command.CreateUserDto
import yu.selab.cj.objectmappingbenchmarkv5.user.command.ReadUserDto
import yu.selab.cj.objectmappingbenchmarkv5.user.entity.User

@Mapper(componentModel = "spring", uses = [PostMapper::class])
interface UserMapper {

    // CreateUserDto -> User 매핑 (posts는 기본값으로 빈 리스트)
    @Mapping(target = "posts", expression = "java(new java.util.ArrayList<Post>())")
    fun toEntity(dto: CreateUserDto): User

    // User -> ReadUserDto 매핑
    // user.uid -> readUserDto.userId
    // user.posts는 PostMapperV5를 통해 List<ReadPostDto>로 변환됨
    @Mapping(source = "uid", target = "userId")
    fun toDto(user: User): ReadUserDto

}